### PR TITLE
Milestone 8: Integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Debas-31**
# Rails Blog App: integration specs for views and fix n+1 problems
>In this milestone of the project implemented the requirements:
- [x] Make sure that the N+1 problem is solved when fetching all posts and their comments for a user. This is the corresponding wireframe.
- [x] Integration specs
     - Use Capybara to write integration tests for each view in your project.
     - User index page:
          - I can see the username of all other users.
          - I can see the profile picture for each user.
          - I can see the number of posts each user has written.
          - When I click on a user, I am redirected to that user's show page.
    - User show page:
         - I can see the user's profile picture.
         - I can see the user's username.
         - I can see the number of posts the user has written.
         - I can see the user's bio.
         - I can see the user's first 3 posts.
         - I can see a button that lets me view all of a user's posts.
         - When I click a user's post, it redirects me to that post's show page.
         - When I click to see all posts, it redirects me to the user's post's index page.
   - User post index page:
        - I can see the user's profile picture.
        - I can see the user's username.
        - I can see the number of posts the user has written.
        - I can see a post's title.
        - I can see some of the post's body.
        - I can see the first comments on a post.
        - I can see how many comments a post has.
        - I can see how many likes a post has.
        - I can see a section for pagination if there are more posts than fit on the view.
        - When I click on a post, it redirects me to that post's show page.
  - Post show page:
       - I can see the post's title.
       - I can see who wrote the post.
       - I can see how many comments it has.
       - I can see how many likes it has.
       - I can see the post body.
       - I can see the username of each commentor.
       - I can see the comment each commentor left.

### CODE REVIEW

File: spec/features/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  it 'Can see the number of posts the user has written.' do
+    expect(page).to have_content 'posts: 1'
+  end
+
+  it 'When I click on a user, I am redirected to that users show page' do
+    click_link(@user.id)
+    expect(current_path).to eq user_path @user.id
+  end
+end
```
Comment (Mihndim2020) [inline spec/features/user_index_spec.rb:28]: - [ ] As it is a  [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md) for this project, "
`Login page:`

  - `I can see the username and password inputs and the "Submit" button.`
  - `When I click the submit button without filling in the username and the password, I get a detailed error.`
  - `When I click the submit button after filling in the username and the password with incorrect data, I get a detailed error.`
  - `When I click the submit button after filling in the username and the password with correct data, I am redirected to the root page.` " , please kindly create a new file and add the above mentioned tests. This [article](https://stackoverflow.com/questions/45198853/testing-sign-up-and-sign-in-with-capybara-rails-rspec-devise) should point you in the right direction.

**Review by Mihndim2020 — CHANGES_REQUESTED** [review]
### Hi @Debas-31, 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Great job with tests coverage
:heavy_check_mark:  N+1 problem well handled
:heavy_check_mark:  Good commit messages
:heavy_check_mark: Descriptive `PR` summary

There are some issues that you still need to work on to get this project approved, but you are almost there!
ALMOST-MINION


# Status: Requires Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

### N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_index_spec.rb (hunk 1 above)
Comment (Debas-31) [reply spec/features/user_index_spec.rb:28]: # Hello, Code reviewers
 In this milestone, we are not supposed to implement the tests for the login page as we are on the new version of the RoR Module, where they missed the order of the milestones. That's why we are including the specs for the four views, not the login. Please get in touch with me on slack for more clarification, reach out to Microverse, or you can refer [to this sneak peek](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/sneak_peek_v1_1.md) of the milestones.

Comment (thecodechaser) [reply spec/features/user_index_spec.rb:28]: @Mihndim2020, The Rails module is updated and the authentication project is after this one. The students don't have login pages yet so I think they can't create tests for the login page. The requirements of this project should be updated. 😄

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @Debas-31👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- No linter errors. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**